File: SQUID_Testing/squidIV.m
function [ptable, Vsquid] = squidIV()

clear all % MATLAB is complaining but this function will only be run like a script
close all
%% Add paths 
addpath('C:\Users\root\Nowack_Lab\Equipment_Drivers');
addpath('C:\Users\root\Nowack_Lab\Utilities');

%% Edit before running

% If testing without a squid, for wiring, etc
no_squid = false;

% Choose parameter file
paramfile = 'std_params_warm3k_davidplay.csv';
parampath = strcat('./Parameters/',paramfile);
[p, ptable] = param_parse(parampath); % use ptable to see parameters in table form

% Git dump? Uncomment if you want a cluttered git.
% git_dump();

%% Define file locations
dropbox = 'C:\Users\root\Dropbox\TeamData\';
time = char(datetime('now','TimeZone','local','Format', 'yyyyMMdd_HHmmss'));

paramsavepath = strcat(dropbox, 'Montana\squid_testing\'); % Where the parameters will be saved
paramsavefile = strcat('squidIV_params_', time, '.csv'); % What the parameters will be called

datapath = strcat(dropbox, 'Montana\squid_testing\'); % Where the data will be saved
datafile = strcat('squidIV_data_', time, '.csv'); % What the data will be called

plotpath = strcat(dropbox, 'Montana\squid_testing\');
plotfile = strcat('squidIV_plot_IV_', time, '.pdf');
plotfile2 = strcat('squidIV_plot_IV_', time, '.png');

%% Ask the user for information
% Check parameters
param_prompt(paramfile);

% Double check no squid
squid_prompt(no_squid);

% Ask for notes
notes = input('Notes about this run: ','s');

%% Some initial checks

% Check for potential SQUIDicide
if ~no_squid
    check_currents(max(abs(p.squid.Irampmax), abs(p.squid.Irampmin)), abs(p.mod.I));
end

% Check to make sure preamp doesn't filter out your signal
if p.preamp.rolloff_high < p.daq.rate
    error('You''re filtering out your signal -____-');
end


%% Send and collect data
nidaq = NI_DAQ(p.daq); % Initializes DAQ parameters
nidaq.set_io('squid'); % For setting input/output channels for measurements done on a SQUID

% Set output data
IsquidR = IVramp(p.squid);
Vmod = p.mod.I * p.mod.Rbias * ones(1,length(IsquidR)); % constant mod current

% prep and send output to the daq
output = [IsquidR; Vmod]; % puts Vsquid into first row and Vmod into second row
[Vsquid, ~] = nidaq.run(output); % Sends a signal to the daq and gets data back
Vsquid = Vsquid/p.preamp.gain; % corrects for gain of preamp

%% Save data, parameters, and notes
data_dump(datafile, datapath,[IsquidR' Vsquid'],{'IsquidR (V)', 'Vsquid (V)'}); % pass cell array to prevent concatentation
copyfile(parampath, strcat(paramsavepath,paramsavefile)); %copies parameter file to permanent location 
fid = fopen(strcat(paramsavepath,paramsavefile), 'a+');
fprintf(fid, '%s', strcat('notes',notes,'none','notes'));
fclose(fid);

%% Plot
figure;
plot_squidIV(gca, IsquidR, Vsquid, p); 
title({['Parameter file: ' paramsavefile];['Data file: ' datafile];['Rate: ' num2str(p.daq.rate) ' Hz']; ['Imod: ' num2str(p.mod.I) ' A']},'Interpreter','none');

print('-dpdf', strcat(plotpath, plotfile));
print('-dpng', strcat(plotpath, plotfile2));

end

File: SQUID_Testing/squidIV.py
# TO DO: 	- Try sweep vs individual points
# 			- Write preamp driver
#			- Instruments as a dict... change this in other code?
#			- Build in acceleration to daq sweep to and from zero?
# 			- Do up and down sweeps?

from IPython import display
import matplotlib.pyplot as plt
import numpy as np

class SquidIV():
    def __init__(self, instruments, squidout, squidin, modout:    
        self.daq = instruments['daq']
        self.preamp = instruments['preamp']
        
        self.squidout = squidout
        self.squidin = squidin
        self.modout = modout
        
        self.savepath = 'C:\Users\Hemlock\Dropbox (Nowack Lab)\TeamData\Montana\squid_testing\'
        
        self.rate = 90 # Hz # measurement rate of the daq
        self.preamp.gain = 500 
        self.preamp.filter = (0, 100) # Hz  
        
        self.Rbias = 2e3 # Ohm # cold bias resistors on the SQUID testing PCB
        self.Rbias_mod = 2e3 # Ohm # cold bias resistors on the SQUID testing PCB
        self.I_mod = 0 # A # constant mod current
        
        self.Irampspan = 200e-6 # A # Will sweep from -Irampspan/2 to +Irampspan/2
		self.Irampstep = 0.5e-6 # A # Step size

		self.numpts = int(self.Irampspan/self.Irampstep)
		
        self.I = np.linspace(-self.Irampspan/2, self.Irampspan/2, self.numpts) # Squid current
        self.Vbias = self.I*self.Rbias # SQUID bias voltage
        self.Vbias_mod = self.I_mod*self.Rbias_mod # Mod bias voltage
        self.V = [] # Measured voltage
        
        self.fig = plt.figure()
        self.ax = plt.gca()
        display.clear_output()
               
    def do(self):
        self.param_prompt() # Check parameters
        setattr(self.daq, self.modout, self.Vbias_mod) # Set mod current
        Vout, Vin, time = self.daq.sweep(self.squidout, -self.Irampspan/2, self.Irampspan/2, freq=self.rate)
        self.V = Vin/self.preamp.gain
        
        self.plot()
        inp = input('Press enter to save data, type anything else to quit. ')
        if inp == '':
        	self.save()
        
        setattr(self.daq, self.modout, 0) # Zero mod current
                          
	def param_prompt(self):
		""" Check and confirm values of parameters """
		for param in ['rate', 'Rbias', 'Rbias_mod', 'Imod', 'Irampspan', 'Irampstep']:
			print(param, ': ', getattr(self, param), '\n')
		for paramamp in ['gain','filter']:
			print('preamp ', param, ': ', getattr(self.preamp, paramamp), '\n')
		correct = False
		while not correct:
			if self.rate > self.preamp.filter_high:
				print("You're filtering out your signal... fix the preamp cutoff\n")
			if self.Irampspan > 200e-6:
				print("You want the SQUID biased above 100 uA?... don't kill the SQUID!\n")
			
			inp = input('Are these parameters correct? Enter a command to change parameters, or press enter to continue (e.g. preamp.gain = 100): ')
			if inp == '':
				correct = True
			else:
				eval('self.'+inp)
         
    def plot(self):
        plt.clf()

        plt.plot(self.I*1e6, self.V, 'k-')
        title('') # NEED DESCRIPTIVE TITLE
        plt.xlabel(r'$I_{bias} = V_{bias}/R_{bias}$ ($\mu$ A)', fontsize=20)
        plt.ylabel(r'$V_{squid}$ (V)', fontsize=20)
                
        display.display(self.fig)